File: frontend/services/animeThemesService.ts

import { SeasonCollection, Song } from "../types";

const ANIMETHEMES_API_URL = 'https://api.animethemes.moe/anime';

interface AnimeThemeArtist {
  name: string;
}

interface AnimeThemeSong {
  title: string;
  artists: AnimeThemeArtist[];
}

interface AnimeThemeVideo {
  basename: string;
  tags: string;
}

interface AnimeThemeEntry {
  videos: AnimeThemeVideo[];
}

interface AnimeTheme {
  type: string; // "OP", "ED", "IN"
  sequence: number | string;
  slug: string;
  song: AnimeThemeSong;
  animethemeentries: AnimeThemeEntry[];
}

interface AnimeSynonym {
  text: string;
}

interface Anime {
  name: string;
  slug: string;
  animethemes: AnimeTheme[];
  animesynonyms?: AnimeSynonym[];
}

interface AnimeThemesResponse {
  anime: Anime[];
}

// Helper to clean string for comparison
const normalizeTokens = (str: string): string[] => {
  return str.toLowerCase()
    .replace(/[^\w\s]/g, ' ') // Replace punctuation with space
    .split(/\s+/)
    .filter(t => t.length > 0);
};

// Check if two titles are relevant matches based on token overlap
const isTitleMatch = (query: string, candidate: string): boolean => {
  const qTokens = normalizeTokens(query);
  const cTokens = normalizeTokens(candidate);
  
  if (qTokens.length === 0 || cTokens.length === 0) return false;

  // Strict check: One set of tokens must be a subset of the other
  const queryInCandidate = qTokens.every(qt => cTokens.includes(qt));
  const candidateInQuery = cTokens.every(ct => qTokens.includes(ct));

  return queryInCandidate || candidateInQuery;
};

export const fetchThemesFromApi = async (animeTitle: string): Promise<SeasonCollection[]> => {
  try {
    // Request: Search for anime, include themes, songs, artists, videos AND synonyms for better matching
    const params = new URLSearchParams({
      'q': animeTitle,
      'include': 'animethemes.song.artists,animethemes.animethemeentries.videos,animesynonyms',
      'limit': '6' 
    });

    const response = await fetch(`${ANIMETHEMES_API_URL}?${params.toString()}`, {
      method: 'GET',
      headers: {
        'Accept': 'application/json',
      },
    });

    if (!response.ok) {
      if (response.status === 404) return [];
      throw new Error(`AnimeThemes REST API Error: ${response.status}`);
    }

    const json = await response.json() as AnimeThemesResponse;
    const rawResults = json.anime || [];

    // Filter results to remove unrelated anime that fuzzy search might have picked up
    const filteredResults = rawResults.filter(anime => {
      // Check main name
      if (isTitleMatch(animeTitle, anime.name)) return true;
      
      // Check synonyms
      if (anime.animesynonyms && anime.animesynonyms.some(s => isTitleMatch(animeTitle, s.text))) {
        return true;
      }
      
      return false;
    });

    if (filteredResults.length === 0) {
      return [];
    }

    // Map the results to our SeasonCollection format
    const collections: SeasonCollection[] = filteredResults.map((anime) => {
      const openings: Song[] = [];
      const endings: Song[] = [];
      const osts: Song[] = [];

      if (!anime.animethemes) return null;

      anime.animethemes.forEach((theme) => {
        const title = theme.song?.title || "Unknown Title";
        const artist = theme.song?.artists?.map(a => a.name).join(", ") || "Unknown Artist";
        
        // Find the best video (prioritize non-creditless if multiple, or just take first)
        const video = theme.animethemeentries?.[0]?.videos?.[0];
        
        // Construct the direct video URL
        const videoUrl = video ? `https://v.animethemes.moe/${video.basename}` : undefined;

        const songObj: Song = {
          title,
          artist,
          videoUrl
        };

        if (theme.type === "OP") {
          openings.push(songObj);
        } else if (theme.type === "ED") {
          endings.push(songObj);
        } else if (theme.type === "IN") {
          // Add Insert songs to OST list
          osts.push(songObj);
        }
      });

      return {
        seasonName: anime.name,
        openings,
        endings,
        osts // Now populated with Insert songs if available
      };
    }).filter((c): c is SeasonCollection => c !== null && (c.openings.length > 0 || c.endings.length > 0 || c.osts.length > 0));

    return collections;

  } catch (error) {
    console.error("AnimeThemes Fetch Error:", error);
    return [];
  }
};

File: backend/services/animethemes_service.py
"""
AnimeThemes service for fetching OP/ED/OST data
Ported from frontend/services/animeThemesService.ts
"""
import logging
import re
from typing import Dict, Any, List
import httpx

logger = logging.getLogger(__name__)

ANIMETHEMES_API_URL = 'https://api.animethemes.moe/anime'

# HTTP timeout configuration to prevent hanging on slow/unresponsive APIs
HTTPX_TIMEOUT = httpx.Timeout(
    connect=5.0,   # 5 seconds to connect
    read=10.0,     # 10 seconds to read response
    write=5.0,     # 5 seconds to write request
    pool=5.0       # 5 seconds to get connection from pool
)


def normalize_tokens(text: str) -> List[str]:
    """
    Helper to clean string for comparison.
    
    Args:
        text: String to normalize
    
    Returns:
        List of normalized tokens
    """
    # Replace punctuation with space
    text = re.sub(r'[^\w\s]', ' ', text.lower())
    # Split and filter empty strings
    return [t for t in text.split() if t]


def is_title_match(query: str, candidate: str) -> bool:
    """
    Check if two titles are relevant matches based on token overlap.
    
    Args:
        query: The search query
        candidate: The candidate title to match against
    
    Returns:
        True if titles match, False otherwise
    """
    q_tokens = normalize_tokens(query)
    c_tokens = normalize_tokens(candidate)
    
    if not q_tokens or not c_tokens:
        return False
    
    # Strict check: One set of tokens must be a subset of the other
    query_in_candidate = all(qt in c_tokens for qt in q_tokens)
    candidate_in_query = all(ct in q_tokens for ct in c_tokens)
    
    return query_in_candidate or candidate_in_query


async def fetch_themes_from_api(anime_title: str) -> List[Dict[str, Any]]:
    """
    Fetches theme data from AnimeThemes API.
    
    Args:
        anime_title: The anime title to search for
    
    Returns:
        List of SeasonCollection dictionaries with openings, endings, and osts
    """
    try:
        # Request: Search for anime, include themes, songs, artists, videos AND synonyms
        params = {
            'q': anime_title,
            'include': 'animethemes.song.artists,animethemes.animethemeentries.videos,animesynonyms',
            'limit': '6'
        }
        
        async with httpx.AsyncClient(timeout=HTTPX_TIMEOUT) as client:
            response = await client.get(
                ANIMETHEMES_API_URL,
                params=params,
                headers={'Accept': 'application/json'}
            )
            
            if response.status_code == 404:
                return []
            
            if not response.is_success:
                raise Exception(f"AnimeThemes REST API Error: {response.status_code}")
            
            data = response.json()
            raw_results = data.get('anime', [])
            
            # Filter results to remove unrelated anime that fuzzy search might have picked up
            filtered_results = []
            for anime in raw_results:
                # Check main name
                if is_title_match(anime_title, anime.get('name', '')):
                    filtered_results.append(anime)
                    continue
                
                # Check synonyms
                synonyms = anime.get('animesynonyms', [])
                if any(is_title_match(anime_title, syn.get('text', '')) for syn in synonyms):
                    filtered_results.append(anime)
            
            if not filtered_results:
                return []
            
            # Map the results to SeasonCollection format
            collections = []
            for anime in filtered_results:
                openings = []
                endings = []
                osts = []
                
                themes = anime.get('animethemes', [])
                if not themes:
                    continue
                
                for theme in themes:
                    song = theme.get('song', {})
                    title = song.get('title', 'Unknown Title')
                    
                    # Get artist names
                    artists = song.get('artists', [])
                    artist = ', '.join([a.get('name', '') for a in artists]) or 'Unknown Artist'
                    
                    # Find the best video
                    entries = theme.get('animethemeentries', [])
                    video = None
                    if entries and entries[0].get('videos'):
                        video = entries[0]['videos'][0]
                    
                    # Construct the direct video URL
                    video_url = f"https://v.animethemes.moe/{video.get('basename')}" if video else None
                    
                    song_obj = {
                        'title': title,
                        'artist': artist,
                        'videoUrl': video_url
                    }
                    
                    theme_type = theme.get('type', '')
                    if theme_type == 'OP':
                        openings.append(song_obj)
                    elif theme_type == 'ED':
                        endings.append(song_obj)
                    elif theme_type == 'IN':
                        # Add Insert songs to OST list
                        osts.append(song_obj)
                
                # Only add if we have at least some themes
                if openings or endings or osts:
                    collections.append({
                        'seasonName': anime.get('name', 'Unknown'),
                        'openings': openings,
                        'endings': endings,
                        'osts': osts
                    })
            
            return collections
            
    except Exception as e:
        logger.error(f"AnimeThemes Fetch Error: {e}")
        return []
